File: test/end-to-end/Fibonacci/C#/Generation/CachingGenerator.cs
//
using Fibonacci.Generation;
using System.Collections.Generic;

namespace Fibonacci.Generation
{

    public class CachingGenerator : IGenerator
    {
        private Dictionary<int, int> cache;

        public CachingGenerator()
        {
            this.cache = new Dictionary<int, int>();
        }

        public int Generate(int index)
        {
            if (index < 2)
            {
                return index;
            }

            int one = this.Generate(index - 1);
            int two = this.Generate(index - 2);
            int result = one + two;

            this.cache[index] = result;

            return result;
        }

        public bool IsCached(int index)
        {
            return this.cache.ContainsKey(index);
        }
    }
}
//

File: test/end-to-end/Fibonacci/Python/generation/caching_generator.py

class CachingGenerator:

    def __init__(self):
        self.__cache = {}

    def generate(self, index):
        if index < 2:
            return index

        one = self.generate(index - 1)
        two = self.generate(index - 2)
        result = one + two

        self.__cache[index] = result

        return result

    def is_cached(self, index):
        return index in self.__cache

